### PR TITLE
[OpPerf] Add norm, cast ops, remaining optimizer ops

### PR DESCRIPTION
**Pull request by ChaiBapchya**
## Description ##

Adds following ops to OpPerf [once all the OpPerf PRs get merged, all MXNet ops in NDArray namespace will be completed in OpPerf coverage (minus deprecated, _contrib ops and few other ops with known issues)]

- norm op to reduction category
- Following Optimizer ops
  - mp_nag_mom
  - nag_mom
  - lambd_update_phase_1 and 2
  - preloaded_multi_*
  - multi_*
- Cast ops
  - cast
  - amp_cast
  - amp_multicast

### Essentials ###
Please feel free to remove inapplicable items for your PR.

- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
Ops falling in the category of multi_* and preloaded_multi_* were notorious
They don't have standard keyworded arguments (data=data)
They expect variable positional args (*data)
It was previously handled such that only 1 value of variable positional arg could be passed

Now, you can pass as many values for positional args as you want.
However for running run_performance_test, inputs have to be key:value pair

That is handled as args0, args1, arg2, etc
So as long as the name of the key starts with args it will be considered as input for the operator.

This is a very rare case that the operator doesn't take keyworded args. Most other ops in MXNet NDArray namespace take keyworded args.

### CODE REVIEW

File: benchmark/opperf/utils/benchmark_utils.py [hunk 1]
```diff
@@ -67,28 +61,17 @@ def _run_nd_operator_performance_test(op, inputs, run_backward, warmup, runs, ar
...
-            _, profiler_output = benchmark_helper_func(op, runs, [], **kwargs)
-
-            # Add inputs used for profiling this operator into result
-            profiler_output["inputs"] = inputs[idx]
-            op_benchmark_result[op.__name__].append(profiler_output)
-    else:
-        for idx, (args, kwargs) in enumerate(zip(args_list, kwargs_list)):
-            _, profiler_output = benchmark_helper_func(op, runs, args, **kwargs)
+    for idx, kwargs in enumerate(kwargs_list):
+        _, profiler_output = benchmark_helper_func(op, runs, [], **kwargs)
```
Comment (connorgoggins) [inline benchmark/opperf/utils/benchmark_utils.py]: We should incorporate our fix (removing the empty list) here as well.

File: benchmark/opperf/utils/ndarray_utils.py [hunk 2]
```diff
@@ -44,10 +42,14 @@ def nd_forward_backward_and_profile(op, runs, *args, **kwargs):
     """
     for _ in range(runs):
         with mx.autograd.record():
-            if not isinstance(args[0], nd.NDArray):
-                res = op(**kwargs)
-            else:
+            args = []
+            for key in kwargs:
+                if key.startswith("args"):
+                    args.append(kwargs.pop(key))
```
Comment (connorgoggins) [inline benchmark/opperf/utils/ndarray_utils.py]: I thought we fixed this?

File: benchmark/opperf/utils/ndarray_utils.py [hunk 3]
```diff
@@ -75,10 +75,14 @@ def nd_forward_and_profile(op, runs, *args, **kwargs):
     any results from NDArray operation execution
     """
     for _ in range(runs):
-        if not isinstance(args[0], nd.NDArray):
-            res = op(**kwargs)
-        else:
+        args = []
+        for key in kwargs:
+            if key.startswith("args"):
+                args.append(kwargs.pop(key))
```
Comment (connorgoggins) [inline benchmark/opperf/utils/ndarray_utils.py]: Same here

File: benchmark/opperf/utils/benchmark_utils.py (hunk 1 above)
Comment (ChaiBapchya) [reply benchmark/opperf/utils/benchmark_utils.py]: Addressed

File: benchmark/opperf/utils/ndarray_utils.py (hunk 2 above)
Comment (ChaiBapchya) [reply benchmark/opperf/utils/ndarray_utils.py]: Addressed

File: benchmark/opperf/utils/ndarray_utils.py (hunk 3 above)
Comment (ChaiBapchya) [reply benchmark/opperf/utils/ndarray_utils.py]: Addressed

**Comment by ChaiBapchya** [comment]
Category specific operators gist : https://gist.github.com/ChaiBapchya/8c4ffb924fe20de1297d53fb4d1a5b1e

Basically result of running
```
>>> from benchmark.opperf.nd_operations.nn_optimizer_operators import run_optimizer_operators_benchmarks
>>> run_optimizer_operators_benchmarks()
```

```
>>> from benchmark.opperf.nd_operations.reduction_operators import run_mx_reduction_operators_benchmarks
>>> run_mx_reduction_operators_benchmarks()
```
```
>>> from benchmark.opperf.nd_operations.unary_operators import run_mx_unary_operators_benchmarks
>>> run_mx_unary_operators_benchmarks()
```

**Comment by ChaiBapchya** [comment]
Entire OpPerf Suite
CPU results : https://gist.github.com/ChaiBapchya/1c26e5a904d9ce9342d61b29f195c5cf

File: benchmark/opperf/rules/default_params.py [hunk 4]
```diff
@@ -31,6 +31,7 @@
 
 # For Unary operators like abs, arccos, arcsin etc..
 DEFAULT_DATA = [(1024, 1024), (10000, 1), (10000, 100)]
+DEFAULT_DTYPE = ['int32'] # required parameter for amp_cast, cast
```
Comment (connorgoggins) [inline benchmark/opperf/rules/default_params.py]: Can we also add `float32` here?

File: benchmark/opperf/utils/benchmark_utils.py [hunk 5]
```diff
@@ -31,28 +31,72 @@
...
+    <NDArray 5x5 @cpu(0)>, 'var':
+    [[ 1.3166187  -0.43292624  0.71535987  0.9254156  -0.90495086]
+     [-0.074684    0.82254    -1.8785107   0.8858836   1.9118724 ]
+     [ 0.33342266  0.11883813 -1.9198899  -0.67558455  1.007749  ]
+     [-0.35391203  1.6323917  -0.33354783 -1.7378405   0.7737382 ]
+     [ 0.89126545  3.2904532  -1.1976235   1.8938874  -0.5669272 ]]
+    <NDArray 5x5 @cpu(0)>, 't': 1, 'wd': 0.1}
+
+    Output
+    {'inputs': {'weight': '<NDArray 5x5 @cpu(0)>', 'grad': '<NDArray 5x5 @cpu(0)>', 'mean': '<NDArray 5x5 @cpu(0)>', 'var': '<NDArray 5x5 @cpu(0)>', 't': 1, 'wd': 0.1}
```
Comment (connorgoggins) [inline benchmark/opperf/utils/benchmark_utils.py]: Nice! Very clean output format.

**Comment by connorgoggins** [comment]
Also can we see updated perf results?

File: benchmark/opperf/rules/default_params.py (hunk 4 above)
Comment (ChaiBapchya) [reply benchmark/opperf/rules/default_params.py]: We can't because cast needs it to be int compulsorily.

**Comment by ChaiBapchya** [comment]
Previously entire NDArray that was input used to get printed
This would not be rendered correctly on Markdown (come up as new lines).
Now after this fix, only the shape of NDArray is stored as a String (much more readable and rendered correctly on Markdown)

CPU : https://gist.github.com/ChaiBapchya/12d7fdd4ac15703e537aefbc8055c981
GPU : https://gist.github.com/ChaiBapchya/c62c7fff4deb115b64ec69b5c4aa6a7d

File: benchmark/opperf/rules/default_params.py (hunk 4 above)
Comment (connorgoggins) [reply benchmark/opperf/rules/default_params.py]: I ran both ops (`cast` and `amp_cast`) with `float32` in the shell and they worked properly.

Comment (ChaiBapchya) [reply benchmark/opperf/rules/default_params.py]: I stand corrected. Fixed it now.

**Comment by ChaiBapchya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by ChaiBapchya** [comment]
Updated OpPerf results after dtype fix
CPU : https://gist.github.com/ChaiBapchya/dbe941cb65d493c6e435c68a6bbc5b19
GPU : https://gist.github.com/ChaiBapchya/479b3b8efb68398d290a5e5863c7809f

**Pull request merged**